MultiLangProgGuide/MultiLangProgGuide.rst

```diff
@@ -899,12 +903,75 @@ ydb_lock_s()
 Note that the parameter list **must** be terminated by a NULL pointer.
 
 Release any locks held by the process, attempt to acquire all the
-specified locks. While the release is unconditional, on return, the
-function will have acquired all specified locks or no specified
-locks. If no locks are specified, the function releases all locks and
-returns ``YDB_OK``.
+requested locks. While the release is unconditional, on return, the
+function will have acquired all requested locks or none of them. If no
+locks are requested, the function releases all locks and returns
+``YDB_OK``.
 
-``timeout`` specifies a time in
+``timeout`` specifies a time in microseconds that the function waits
+to acquire the requested locks. If it is not able to acquire all
+requested locks, it acquires no locks, returning with a
+``YDB_LOCK_TIMEOUT`` return value.
+
+If ``timeout`` is zero, the function makes exactly one attempt to
+acquire the locks, and if it is unable to, it returns
+``YDB_LOCK_TIMEOUT``.
+
+If all requested locks are successfully acquired, the function returns
+``YDB_OK``.
+
+-----------------
+ydb_lock_decr_s()
+-----------------
+
+.. code-block:: C
+
+	int ydb_lock_s([int count,
+		ydb_string_t *varname[,
+		ydb_string_t *subscript, ...], ...,] NULL);
+
+Note that the parameter list **must** be terminated by a NULL pointer.
```

Comment:
Why? What's the purpose of count then? I'd rather stick with count and not have a NULL terminated list.

Reply:
The argument list can consist of multiple lock resource names, each of which can have multiple subscripts. Each resource name has a count of the number of subscripts, and the list of resources is terminated with a NULL.

Reply:
This is a very unwieldy and difficult-to-use thing. It is also heavy weight. If someone is grabbing 2 or 3 locks with 3 or 4 parms each, that's a large amount of parameter processing and housekeeping on every call. It cries for something somewhat simpler both for the sake of ease of use and for performance.

One thought I had was to make it slightly more structured (pardon the pun) - perhaps if only one lock is wanted, we could use the simple parameter list to specify the var name and its subscripts if any (similar to get/put) but if multiple locks were to be grabbed/released, Use an alternate interface (multilock) using a structure something like the following:

typedef struct
{
        int                     count;    /* Count of subscripts */
        gtm_buffer_t     varname;
        gtm_buffer_t     subscripts[1];
} ydb_lock_t;

This block would describe a lock. Note the block would be allocated with "count" subscripts so this is a variable length block. Could even have a macro that instantiates this block with a specific value for subscripts so it is a fixed size block. Several possibilities here including have the varname and subscripts be addresses instead of fixed blocks but I think the fixed blocks is fine for a first shot at this and getting some reactions.

With this structure defined and filled in for each lock (I envision these mostly being stack allocated structures), we can simplify a multi-lock scenario to a count and a list of addresses of the ydb_lock_t structures. No need for trailing NULLs to terminate a list.

Reply:
As discussed, the parameters will have a count of the number of resources rather than a NULL at the end.